### ISSUE TITLE
FR: Add Norway, Sweden, Finland and Germany zones

### ISSUE THREAD

**Issue by MTrab** [MTrab/energidataservice#16]
Add the remaining zones (Norway, Sweden, Finland and Germany) as valid selections

**Comment by mikael-andersson** [comment]
I would like to add Swedish time zones but need some help.

I have identified the following places where I guess I need to add more price areas:

# sensor.py

```
            if curentity.unique_id in [
                "energidataservice_West of the great belt",
                "energidataservice_East of the great belt",
            ]:
```

# const.py

```
AREA_DK_EAST = "East of the great belt"
AREA_DK_WEST = "West of the great belt"
```

```
# Mappings - OLD STYLE
AREA_MAP = {  # Text-to-area
    AREA_DK_WEST: "DK1",
    AREA_DK_EAST: "DK2",
}
AREA_TO_TEXT = {  # Area-to-text
    "DK1": AREA_DK_WEST,
    "DK2": AREA_DK_EAST,
}
```

When I add a new price area in above places, the price area pops up when configuring the integration, but I do not get any sensor. Any ideas?

**Comment by MTrab** [comment]
Current release can ONLY be used for the 2 regions in Denmark, it's not compatible with others. That's the reason for this FR.
Currently working on refactoring the integration, so it becomes compatible with every zone available at Energi Data Service

**Comment by mikael-andersson** [comment]
Thank you so much for the response. That is great you are working on it! :) Do you have a forecast, is it about a month away? I was not able to see anything else specific for the Danish price areas in the code. Where in the code do you need to make the proper changes? Is there something I can help out with?

**Comment by MTrab** [comment]
No - don't have an ETA on when it'll be expected to be released.
Weather have been to good to sit inside and be coding for the last few weeks ;)

**Comment by mikael-andersson** [comment]
No problem if you do not have an ETA, just wondered if you had a rough estimate, like if it was a month, a year or a decade. :)

Sounds very good. :)

**Comment by MTrab** [comment]
Hmmm - this is not promising news.
Energi Data Service only guarantees data for DK1 and DK2.
A few other zones are available too, but not all.
IE. today only SE3 and SE4 zones are available, but not SE1 and SE2.
As well as alot of Norwegian zones are missing too.

**Comment by mikael-andersson** [comment]
Yes, it is a pity. But do you/we really need to use Energi Data Service API, why not use ENTSO-E directly since that is what Energi Data Service is using, is it not?

I should say that I have neither looked at ENTSO-E's API nor Energi Data Service so I do not know the differences.

**Comment by MTrab** [comment]
Actually working, right now, on making it possible for this integration to pull data from multiple sources and use fallback methods if one doesn't provide any data

**Comment by MTrab** [comment]
And can't seem to find anything on ENTSO-E regarding spot prices or API in general.

**Comment by mikael-andersson** [comment]
Hmm, I do not know where I got it from that it was based on ENTSO-E's API... I cannot find anything now.

Anyhow. have you considered using ENTSO-E (https://www.entsoe.eu/)? See e.g. https://pypi.org/project/entsoe-py/

Here is some data for SE3 in Sweden:
https://transparency.entsoe.eu/transmission-domain/r2/dayAheadPrices/show?name=&defaultValue=false&viewType=GRAPH&areaType=BZN&atch=false&dateTime.dateTime=26.04.2022+00:00|CET|DAY&biddingZone.values=CTY|10YSE-1--------K!BZN|10Y1001A1001A46L&resolution.values=PT15M&resolution.values=PT30M&resolution.values=PT60M&dateTime.timezone=CET_CEST&dateTime.timezone_input=CET+(UTC+1)+/+CEST+(UTC+2)

I have an account at ENTSO-E to be able to use the API but I have not tried it out yet.

**Comment by mikael-andersson** [comment]
It is true, either you would have to add your personal API key to your integration or you have to setup a similar service as EnergiDataService. I actually do not know if one is allowed to do that with ENTSO-E. I think it is fine but that is something one needs to check with ENTSO-E.

The big difference between EnergiDataService and ENTSO-E is, I suppose, that EnergiDataService, provides different currencies while ENTSO-E only has EUR.

Have you contacted EnergiDataService if they are working on releasing more zones/price areas? If they have e.g. SE3 and SE4 and would be a piece of cake to extend with the others in Sweden.

**Comment by mikael-andersson** [comment]
Ok, then I understand. Yes, maybe they are not interested at all to extend with more zones then.

I am very interested in an integration based on ENTSO-E anyhow but the API key makes it a bit more complicated.

As for currencies, maybe I was wrong, does not EnergiDataService provide the price in the right currency or is that something you handle in this integration?

**Comment by MTrab** [comment]
Currency conversion is handled in the integration, based on you local Home Assistant currency settings.

**Comment by MTrab** [comment]
No, currently I only pull EUR prices anyway

**Comment by mikael-andersson** [comment]
Ok.

**Comment by MTrab** [comment]
@mikael-andersson if you haven't seen it already, I have published a [wiki page](https://github.com/MTrab/energidataservice/wiki/Add-new-custom-API-connector) where I try to describe how to add custom connectors in the upcoming release

**Comment by mikael-andersson** [comment]
Great, thanks for the information.
